File: Windows/Domain/ResearchModels.cs
using System.Text.Json.Serialization;

namespace CytisusTrading.Windows;

[JsonConverter(typeof(JsonStringEnumConverter))]
public enum FactorTaskType
{
    Alpha,
    Risk,
    Regime,
    Liquidity,
    Execution
}

[JsonConverter(typeof(JsonStringEnumConverter))]
public enum FactorTrialResult
{
    Candidate,
    Rejected,
    Quarantined
}

[JsonConverter(typeof(JsonStringEnumConverter))]
public enum FactorMissingValuePolicy
{
    Propagate,
    IgnoreWindowMissing,
    CrossSectionMedian
}

public static class FactorHorizons
{
    public static IReadOnlyList<string> Daily { get; } =
        new[] { "1d", "3d", "5d", "10d", "20d", "60d", "120d", "252d" };

    public static IReadOnlySet<string> ReservedIntraday { get; } =
        new HashSet<string>(StringComparer.Ordinal)
        {
            "1m",
            "5m",
            "30m"
        };

    public static int Days(string horizon)
    {
        if (!Daily.Contains(horizon, StringComparer.Ordinal))
        {
            throw new ArgumentOutOfRangeException(
                nameof(horizon),
                "Only standard daily horizons are evaluated.");
        }
        return int.Parse(
            horizon[..^1],
            System.Globalization.CultureInfo.InvariantCulture);
    }
}

public sealed record FactorDefinition(
    int SchemaVersion,
    string FactorId,
    FactorTaskType FactorType,
    string Target,
    string Horizon,
    string Universe,
    IReadOnlyList<string> RequiredData,
    string UpdateFrequency,
    int Version,
    FactorState State,
    string Expression,
    IReadOnlyDictionary<string, FactorState> StrategyStates);

public sealed record FactorEvidence(
    double Coverage,
    double MissingRate,
    double TaskMetric,
    string TaskMetricName,
    double Stability,
    double Turnover,
    double CostProxy,
    double CapacityProxy,
    double ExistingFactorCorrelation,
    double NeighboringHorizonStability,
    double MarginalContribution,
    int OosWindows,
    double MultipleTestingPenalty,
    string RegimeEvidence);

public sealed record FactorTrial(
    int SchemaVersion,
    string TrialId,
    string FactorFamily,
    string Expression,
    IReadOnlyDictionary<string, string> Parameters,
    string DatasetVersion,
    string UniverseVersion,
    string SearchAlgorithm,
    int Generation,
    int OosWindows,
    IReadOnlyDictionary<string, double> Metrics,
    FactorTrialResult Result,
    string? RejectionReason,
    DateTimeOffset CreatedAt);

public sealed record FactorResearchSummary(
    FactorDefinition Definition,
    FactorState GlobalState,
    FactorState StrategyState,
    int TrialCount,
    FactorEvidence Evidence,
    string StateReason)
{
    public string FactorId => Definition.FactorId;
    public string Name => Definition.FactorId
        .Replace("-", " ", StringComparison.Ordinal);
    public string FactorType => Definition.FactorType.ToString();
    public string Horizon => Definition.Horizon;
    public string GlobalStateLabel => GlobalState.ToString();
    public string StrategyStateLabel => StrategyState.ToString();
    public string OosEvidence =>
        $"{Evidence.OosWindows} windows | {Evidence.TaskMetricName} {Evidence.TaskMetric:0.000}";
    public string RegimeEvidence => Evidence.RegimeEvidence;
    public string MarginalContribution =>
        Evidence.MarginalContribution.ToString(
            "0.000",
            System.Globalization.CultureInfo.InvariantCulture);
}

public sealed record ResearchObservation(
    string Symbol,
    string Industry,
    DateTimeOffset EventTime,
    DateTimeOffset AvailableTime,
    double Open,
    double High,
    double Low,
    double Close,
    double Volume,
    double Returns,
    double MarketReturn,
    double SectorReturn,
    double CapitalFlow)
{
    public double Field(string name)
    {
        return name switch
        {
            "open" => Open,
            "high" => High,
            "low" => Low,
            "close" => Close,
            "volume" => Volume,
            "returns" => Returns,
            "market_return" => MarketReturn,
            "sector_return" => SectorReturn,
            "capital_flow" => CapitalFlow,
            _ => throw new ArgumentOutOfRangeException(
                nameof(name),
                "Unknown research field.")
        };
    }
}

public sealed record NormalizedResearchDataset(
    string DatasetVersion,
    string UniverseVersion,
    DateTimeOffset CreatedAt,
    IReadOnlyList<ResearchObservation> Observations);

public sealed record FactorSearchConfiguration(
    int BeamWidth,
    int TopK,
    int MaximumGenerations,
    int MaximumCandidates,
    int MaximumDepth,
    int MaximumOperators)
{
    public static FactorSearchConfiguration Tiny { get; } =
        new(4, 3, 2, 24, 8, 16);
}

public sealed record FactorSearchResult(
    string Status,
    int EvaluatedCount,
    int CandidateCount,
    int RejectedCount,
    int QuarantinedCount,
    IReadOnlyList<FactorTrial> Trials,
    IReadOnlyList<FactorResearchSummary> Summaries);

public sealed record FactorLifecycleEvidence(
    bool DataContamination,
    bool LookAheadBias,
    bool TrainTestLeakage,
    bool UnreproducibleResult,
    bool DefinitionError,
    bool TrialRecordPresent,
    double ShortTermScore,
    double LongTermScore,
    bool RegimeSpecificFailure,
    int RepeatedOosFailures,
    double ChangePointProbability,
    double CounterfactualImprovement,
    bool RenewedEvidence);

public sealed record FactorLifecycleDecision(
    FactorState State,
    bool Global,
    bool BlocksRiskExpansion,
    double RiskMultiplier,
    string Reason);

public sealed record RegimeInput(
    double TrendStrength,
    double RangeScore,
    double RealizedVolatility,
    double CrossAssetRisk,
    IReadOnlyDictionary<string, double> StrategyReportedFit);

public sealed record RegimeSnapshot(
    int SchemaVersion,
    DateTimeOffset AsOf,
    IReadOnlyDictionary<string, double> Probabilities,
    double Uncertainty,
    double RiskMultiplier,
    IReadOnlyDictionary<string, double> Explanation)
{
    public double Trend => Probabilities.GetValueOrDefault("Trend");
    public double Range => Probabilities.GetValueOrDefault("Range");
    public double HighVolatility =>
        Probabilities.GetValueOrDefault("HighVolatility");
    public double Crisis => Probabilities.GetValueOrDefault("Crisis");
}

public sealed record StrategyAllocationInput(
    string StrategyId,
    HealthState Health,
    bool IsNew,
    double LongTermOos,
    double RecentLive,
    double RecentPaper,
    double SignalStrength,
    double RegimeFit,
    double ConfidenceCalibration,
    double DataQuality,
    double ModelUncertainty,
    double Volatility,
    double AverageCorrelation,
    double Drawdown,
    double Capacity,
    double Liquidity,
    double CurrentWeight);

public sealed record RegimeAllocationFixture(
    int SchemaVersion,
    RegimeInput RegimeInput,
    double TotalCapital,
    double BaseRiskFraction,
    IReadOnlyList<StrategyAllocationInput> Strategies);

public sealed record StrategyCapitalAllocation(
    int SchemaVersion,
    string StrategyId,
    double FinalWeight,
    double CapitalBudget,
    IReadOnlyDictionary<string, double> Explanation)
{
    public string ExplanationText =>
        $"Base {Explanation.GetValueOrDefault("base_weight"):P1} | " +
        $"Alpha {Explanation.GetValueOrDefault("alpha_tilt"):+0.000;-0.000;0.000} | " +
        $"Correlation {Explanation.GetValueOrDefault("correlation_penalty"):0.000} | " +
        $"Drawdown {Explanation.GetValueOrDefault("drawdown_penalty"):0.000} | " +
        $"Capacity {Explanation.GetValueOrDefault("capacity_penalty"):0.000} | " +
        $"Liquidity {Explanation.GetValueOrDefault("liquidity_penalty"):0.000}";
}

public sealed record CapitalAllocationResult(
    double TotalRiskBudget,
    RegimeSnapshot Regime,
    IReadOnlyList<StrategyCapitalAllocation> Allocations);

File: Sources/Domain/ResearchModels.swift
import Foundation

enum FactorTaskType: String, Codable, CaseIterable {
    case alpha = "Alpha"
    case risk = "Risk"
    case regime = "Regime"
    case liquidity = "Liquidity"
    case execution = "Execution"
}

enum FactorTrialResult: String, Codable {
    case candidate = "Candidate"
    case rejected = "Rejected"
    case quarantined = "Quarantined"
}

enum FactorMissingValuePolicy: String, Codable {
    case propagate = "Propagate"
    case ignoreWindowMissing = "IgnoreWindowMissing"
    case crossSectionMedian = "CrossSectionMedian"
}

enum FactorHorizons {
    static let daily = ["1d", "3d", "5d", "10d", "20d", "60d", "120d", "252d"]
    static let reservedIntraday = Set(["1m", "5m", "30m"])

    static func days(_ horizon: String) throws -> Int {
        guard daily.contains(horizon),
              let value = Int(horizon.dropLast()) else {
            throw FactorDSLError.invalidDefinition(
                "Only standard daily horizons are evaluated."
            )
        }
        return value
    }
}

struct FactorDefinition: Codable, Identifiable, Equatable {
    let schemaVersion: Int
    let factorId: String
    let factorType: FactorTaskType
    let target: String
    let horizon: String
    let universe: String
    let requiredData: [String]
    let updateFrequency: String
    let version: Int
    let state: FactorState
    let expression: String
    let strategyStates: [String: FactorState]

    var id: String { factorId }
}

struct FactorEvidence: Codable, Equatable {
    let coverage: Double
    let missingRate: Double
    let taskMetric: Double
    let taskMetricName: String
    let stability: Double
    let turnover: Double
    let costProxy: Double
    let capacityProxy: Double
    let existingFactorCorrelation: Double
    let neighboringHorizonStability: Double
    let marginalContribution: Double
    let oosWindows: Int
    let multipleTestingPenalty: Double
    let regimeEvidence: String
}

struct FactorTrial: Codable, Identifiable, Equatable {
    let schemaVersion: Int
    let trialId: String
    let factorFamily: String
    let expression: String
    let parameters: [String: String]
    let datasetVersion: String
    let universeVersion: String
    let searchAlgorithm: String
    let generation: Int
    let oosWindows: Int
    let metrics: [String: Double]
    let result: FactorTrialResult
    let rejectionReason: String?
    let createdAt: Date

    var id: String { trialId }
}

struct FactorResearchSummary: Identifiable, Equatable {
    let definition: FactorDefinition
    let globalState: FactorState
    let strategyState: FactorState
    let trialCount: Int
    let evidence: FactorEvidence
    let stateReason: String

    var id: String { definition.factorId }
    var name: String {
        definition.factorId.replacingOccurrences(of: "-", with: " ")
    }
    var oosEvidence: String {
        "\(evidence.oosWindows) windows | \(evidence.taskMetricName) " +
            String(format: "%.3f", evidence.taskMetric)
    }
}

struct ResearchObservation: Equatable {
    let symbol: String
    let industry: String
    let eventTime: Date
    let availableTime: Date
    let open: Double
    let high: Double
    let low: Double
    let close: Double
    let volume: Double
    let returns: Double
    let marketReturn: Double
    let sectorReturn: Double
    let capitalFlow: Double

    func field(_ name: String) throws -> Double {
        switch name {
        case "open": return open
        case "high": return high
        case "low": return low
        case "close": return close
        case "volume": return volume
        case "returns": return returns
        case "market_return": return marketReturn
        case "sector_return": return sectorReturn
        case "capital_flow": return capitalFlow
        default:
            throw FactorDSLError.invalidDefinition(
                "Unknown research field: \(name)"
            )
        }
    }
}

struct NormalizedResearchDataset: Equatable {
    let datasetVersion: String
    let universeVersion: String
    let createdAt: Date
    let observations: [ResearchObservation]
}

struct FactorSearchConfiguration: Equatable {
    let beamWidth: Int
    let topK: Int
    let maximumGenerations: Int
    let maximumCandidates: Int
    let maximumDepth: Int
    let maximumOperators: Int

    static let tiny = FactorSearchConfiguration(
        beamWidth: 4,
        topK: 3,
        maximumGenerations: 2,
        maximumCandidates: 24,
        maximumDepth: 8,
        maximumOperators: 16
    )
}

struct FactorSearchResult {
    let status: String
    let evaluatedCount: Int
    let candidateCount: Int
    let rejectedCount: Int
    let quarantinedCount: Int
    let trials: [FactorTrial]
    let summaries: [FactorResearchSummary]
}

struct FactorLifecycleEvidence {
    let dataContamination: Bool
    let lookAheadBias: Bool
    let trainTestLeakage: Bool
    let unreproducibleResult: Bool
    let definitionError: Bool
    let trialRecordPresent: Bool
    let shortTermScore: Double
    let longTermScore: Double
    let regimeSpecificFailure: Bool
    let repeatedOosFailures: Int
    let changePointProbability: Double
    let counterfactualImprovement: Double
    let renewedEvidence: Bool
}

struct FactorLifecycleDecision {
    let state: FactorState
    let global: Bool
    let blocksRiskExpansion: Bool
    let riskMultiplier: Double
    let reason: String
}

struct RegimeInput: Codable, Equatable {
    let trendStrength: Double
    let rangeScore: Double
    let realizedVolatility: Double
    let crossAssetRisk: Double
    let strategyReportedFit: [String: Double]
}

struct RegimeSnapshot: Codable, Equatable {
    let schemaVersion: Int
    let asOf: Date
    let probabilities: [String: Double]
    let uncertainty: Double
    let riskMultiplier: Double
    let explanation: [String: Double]

    var trend: Double { probabilities["Trend"] ?? 0 }
    var range: Double { probabilities["Range"] ?? 0 }
    var highVolatility: Double {
        probabilities["HighVolatility"] ?? 0
    }
    var crisis: Double { probabilities["Crisis"] ?? 0 }
}

struct StrategyAllocationInput: Codable, Equatable {
    let strategyId: String
    let health: HealthState
    let isNew: Bool
    let longTermOos: Double
    let recentLive: Double
    let recentPaper: Double
    let signalStrength: Double
    let regimeFit: Double
    let confidenceCalibration: Double
    let dataQuality: Double
    let modelUncertainty: Double
    let volatility: Double
    let averageCorrelation: Double
    let drawdown: Double
    let capacity: Double
    let liquidity: Double
    let currentWeight: Double
}

struct RegimeAllocationFixture: Codable, Equatable {
    let schemaVersion: Int
    let regimeInput: RegimeInput
    let totalCapital: Double
    let baseRiskFraction: Double
    let strategies: [StrategyAllocationInput]
}

struct StrategyCapitalAllocation: Codable, Identifiable, Equatable {
    let schemaVersion: Int
    let strategyId: String
    let finalWeight: Double
    let capitalBudget: Double
    let explanation: [String: Double]

    var id: String { strategyId }
    var explanationText: String {
        let base = explanation["base_weight"] ?? 0
        let alpha = explanation["alpha_tilt"] ?? 0
        let correlation = explanation["correlation_penalty"] ?? 0
        let drawdown = explanation["drawdown_penalty"] ?? 0
        let capacity = explanation["capacity_penalty"] ?? 0
        let liquidity = explanation["liquidity_penalty"] ?? 0
        return String(
            format: "Base %.1f%% | Alpha %+.3f | Correlation %.3f | Drawdown %.3f | Capacity %.3f | Liquidity %.3f",
            base * 100,
            alpha,
            correlation,
            drawdown,
            capacity,
            liquidity
        )
    }
}

struct CapitalAllocationResult: Equatable {
    let totalRiskBudget: Double
    let regime: RegimeSnapshot
    let allocations: [StrategyCapitalAllocation]
}
